### PR TITLE
Session integration + new query builder

### LINKED ISSUE #23: Port Conflicts/Proxying
Our development environment uses two servers - one to serve the react client (a webpack dev server) and our backend (an express server). Both the client and server are currently configured to run on port 3000. As such, the two servers conflict with each other when you try to run both at the same time. In addition, the client is going to be making same-site requests which will currently will go to the webpack dev server instead of our backend.

Firstly we need to reconfigure the port of our express server to be different from the client. I recommend changing the [default configuration value](https://github.com/CS320S22T3/taskbox/blob/40e3f3a2abdc89eae04cba90e46a4e2350d7d01c/server/.env.local.example#L9) to  `3201`. Since it's just a template, everyone on the team will have to update their own `.env` file accordingly.

Secondly we need to configure a proxy from the webpack dev server to the express server. The process for doing that is laid out by [this guide](https://create-react-app.dev/docs/proxying-api-requests-in-development/). To be clear, you want to target `client/package.json`.

### PR DESCRIPTION
**Pull request by daniel-melanson**
Will Resolve #23 #25 #28 #29 #30 #31 

- [x] Port Conflict
- [x] Fix `delete /api/sessions` hang
- [x] Express session secret
- [x] audit npm packages
- [x] Use knex
- [x] Integrating sessions API
- [x] PG Tools

### CODE REVIEW

**Review by absmith22 — DISMISSED** [review]
This looks good. Knex is looking much cleaner than Sequelize too!

File: server/src/routes/sessions.route.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      if (user && bcrypt.compareSync(password, user.password_digest)) {
+        req.session.user_id = String(user.id);
+        return res.status(200).json({ user_id: user.id });
+      } else {
+        return res.status(422).json({ error: "Incorrect email or password." });
+      }
+    } catch (e) {
+      console.log("Error while trying to find user: " + e);
+
+      return res.sendStatus(500);
```
Comment (benmelz) [inline server/src/routes/sessions.route.ts:65]: 404 might be appropriate

File: server/src/routes/validate.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    await Promise.all(validations.map((validation) => validation.run(req)));
+
+    const errors = validationResult(req);
+    if (errors.isEmpty()) {
+      return next();
+    }
+
+    res.status(400).json({ errors: errors.array() });
+  };
+}
```
Comment (benmelz) [inline server/src/routes/validate.ts:19]: This is awesome - my one comment is that it looks to me like it qualifies as middleware. As such you might consider sticking it under a /middleware folder

File: server/db/migrations/20220328193739_init.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  });
+};
+
+/**
+ * @param { import("knex").Knex } knex
+ * @returns { Promise<void> }
+ */
+exports.down = function (knex) {
+  knex.schema.dropTable("users");
+};
```
Comment (benmelz) [inline server/db/migrations/20220328193739_init.js:19]: This should still be named `_create_users.js` - the main idea behind migrations is that they build database schemas iteratively and in increments (once merged this migration will effectively be "finalized" and shouldn't be edited, so it's convention to name the file according to what it does).

**Review by matt532 — APPROVED** [review]

**Review by jeetmitra01 — APPROVED** [review]
Was able to log in and log out successfully

**Review by absmith22 — APPROVED** [review]
Looked over the updates from Ben's suggestions.

**Pull request merged**